### PR TITLE
Fix DNF5 repoquery

### PR DESCRIPTION
**Pull request by alimirjamali**
related: https://github.com/QubesOS/qubes-core-admin-client/pull/308

### CODE REVIEW

**Comment by marmarek** [comment]
Downloading templates fail. I guess "downloadurl" plugin is no longer compatible?

**Comment by alimirjamali** [comment]
> Downloading templates fail. I guess "downloadurl" plugin is no longer compatible?

I am not sure. I am experiencing other bugs. With my previous patch (using `%{reponame}` and allowing space in Regex), the `qvm-template-gui` was working fine. It is now broken. I have to investigate.

**Comment by alimirjamali** [comment]
I got that mixed a little.

We have two problems:
> Downloading templates fail. I guess "downloadurl" plugin is no longer compatible?
1. Yes. downloadurl is no longer available. What to do now? Revert to `dnf repoquery package --location` or `dnf download package --url`?

2. The 2nd issue is that dnf5 does not add `\n` to `dnf repoquery --qf=...`.

**Comment by alimirjamali** [comment]
@marmarek 

The patch that I just uploaded should work for both DNF4 and DNF5. But it is ugly and i do not like it. The problem with skipping `downloadurl` is that all downloads will be from one mirror in R4.3 (and Fedora 41 based updatevms). And  I did not manage to find any way to get a list of all urls to shuffle.

**Comment by alimirjamali** [comment]
Created an Issue on upstream:

https://github.com/rpm-software-management/dnf5/issues/1734

**Comment by alimirjamali** [comment]
[This line](https://github.com/rpm-software-management/dnf5/blob/main/dnf5/commands/download/download.cpp#L276) in dnf5 download plugin is what that has to be changed/improved to allow `--all-mirrors`. I am not sure if the upstream DNF5 developers would accept a `--all-mirrors` option. If not, it should be possible to write a fork of dnf5 download plugin to do the same as downloadurl fork of dnf4's download plugin.

In the meantime, the ugly patch I suggested in this PR should allow temporal use of `qvm-template` with Fedora 41 based updatevms.

@marmarek What is your suggestion?

**Comment by marmarek** [comment]
Well, first mirror is better than qvm-template broken completely. But that isn't suitable long term solution, without listing more URLs, fallback to other mirrors won't work (and it is needed sometimes, especially when downloading over Tor).

**Comment by marmarek** [comment]
What about adding a DNF5 "downloadurl" plugin, based on the "download" command upstream? It was pretty simple to make "downloadurl" plugin for DNF4 ;)

**Comment by marmarek** [comment]
In the meantime, I'll merge some of the other PRs and build initial template for testing. It will have broken VM-side of qvm-template still, but will ease testing other stuff (and also allows running this PR on openQA)

**Comment by alimirjamali** [comment]
Update: Path to `download.cpp` and `download.hpp` is done. I am trying to summit the [patch](https://github.com/rpm-software-management/dnf5/pull/1735) to upstream. Maybe DNF5 maintainers would accept it and other people could use it as well.

**Comment by alimirjamali** [comment]
Update: Individual `downloadurl` plugin for DNF5 is ready and tested.

In the next step, we have to decide how long to wait for the upstream maintainers to make a decision if they like the patch or reject it. If they reject it or do not make a decision within the reasonable time, packaging & building instructions for Qubes OS specific plugin is available [here](https://dnf5.readthedocs.io/en/latest/tutorial/plugins/dnf5-plugins.html).

**Comment by marmarek** [comment]
> OpenQA test summary

The qvm-template related tests seems to be green :) 
But there are a few failures related to qubes-vm-update (but that's unrelated to this PR)

**Comment by alimirjamali** [comment]
> > OpenQA test summary
> 
> The qvm-template related tests seems to be green :) (but that's ofc without the --all-mirrors support in DNF5 yet) But there are a few failures related to qubes-vm-update (but that's unrelated to this PR)

That is good. In the meantime, DNF maintainers are in the process of approving the new option. And they asked it to be renamed to `--allmirrors`. And it requires `-q` to filter out repo refreshing info. But overall I guess it would eliminate the necessity for a Qubes specific plugin.

**Comment by alimirjamali** [comment]
> (but that's ofc without the --all-mirrors support in DNF5 yet)

OK. Upstream DNF5 maintainers have accepted [`--allmirrors` PR](https://github.com/rpm-software-management/dnf5/pull/1735) and merged it. It appears that they follow a (bi)weekly new release schedule. I guess it should be readily available very soon.

On the issue of `dnf repoquery --qf=...` not printing `\n` on DNF5, I am not certain if this is a new deliberate feature or a bug. Overall it is not very serious.

**Comment by marmarek** [comment]
> OK. Upstream DNF5 maintainers have accepted [`--allmirrors` PR](https://github.com/rpm-software-management/dnf5/pull/1735) and merged it. It appears that they follow a (bi)weekly new release schedule. I guess it should be readily available very soon.

Can you adjust this PR to work with the updated dnf5 already, including detection if the `--allmirrors` option is there? That would mean 3 cases now:
- dnf4 - custom plugin
- old dnf5 - take just the first url
- new dnf5 - use `--allmirrors` option

**Comment by alimirjamali** [comment]
> Can you adjust this PR to work with the updated dnf5 already,

Of course I can. But I was under the impression that we should wait until DNF5 maintainers make a new release which includes the `--allmirror` patch? The current DNF5 version on Fedora 41 and Fedora Rawhide is 5.2.6.2 from 3 weeks ago which does not include the  `--allmirror` patch.

https://packages.fedoraproject.org/pkgs/dnf5/dnf5/

**Comment by marmarek** [comment]
Given F41 enters final freeze in 4 days, I'm skeptical this update will make it to the F41 release on day 0, more likely it will be an update later...

**Comment by alimirjamali** [comment]
> Given F41 enters final freeze in 4 days, I'm skeptical this update will make it to the F41 release on day 0, more likely it will be an update later...

I rushed in to replying to your previous message after reading the 1st line  (sorry for that. some lack of sleep here). I just adjusted the PR. I have not tested it personally as I am not currently behind my Qubes OS machine. Hopefully the adjustment should work without any issues as it quite simple. I appreciate if you could do a sanity check.

**Comment by alimirjamali** [comment]
This will not work. DNF5 maintainers had asked me for space-separated URLs rather than each on a new line. readarray has to be adjusted as well.

**Comment by alimirjamali** [comment]
@marmarek 
The patch is now ready. I had to readjust it since you made a change last week which resulted to a merge conflict.

File: qubes-rpc/qvm-template-repo-query
Comment (marmarek) [inline qubes-rpc/qvm-template-repo-query:71]: Does it make sens if you assign plain string to `urls` below anyway? And then `readarray` will create array later. Only the middle case benefits from this declare, but all cases should be consistent.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: qubes-rpc/qvm-template-repo-query
Comment (alimirjamali) [reply qubes-rpc/qvm-template-repo-query:71]: Then I guess the 1st case should use something else for temporary storage of space-separated urls. I will amend it and push the change in a second.

File: qubes-rpc/qvm-template-repo-query
Comment (marmarek) [inline qubes-rpc/qvm-template-repo-query:81]: Maybe you can simply change this to `urls=($url)`? I don't think there is anything in the `urls` at this point yet?

Comment (alimirjamali) [reply qubes-rpc/qvm-template-repo-query:81]: OK. Will do it as well

**Comment by alimirjamali** [comment]
@marmarek The two requested changes are applied.

**Review by marmarek — APPROVED** [review]

**Pull request merged**
